File: javaAudioCryption/encode.java
import javax.sound.sampled.*;
import java.io.*;
import java.nio.charset.StandardCharsets;

public class encode {

    public static void hideMessageInWav(String inputWavPath, String messageFilePath, String outputWavPath) {
        try {

            AudioInputStream waveRead = AudioSystem.getAudioInputStream(new File(inputWavPath));
            AudioFormat format = waveRead.getFormat();
            byte[] frames = new byte[(int) waveRead.getFrameLength() * format.getFrameSize()];
            waveRead.read(frames);


            messageFilePath += "=====";
            byte[] messageBytes = messageFilePath.getBytes(StandardCharsets.UTF_8);

            if (messageBytes.length * 8 > frames.length) {
                throw new IllegalArgumentException("Сообщение слишком длинное для этого аудиофайла.");
            }

            int messageBitIndex = 0;
            for (byte b : messageBytes) {
                for (int j = 0; j < 8; j++) {
                    int bit = (b >> (7 - j)) & 1;
                    frames[messageBitIndex] = (byte) ((frames[messageBitIndex] & 0xFE) | bit);
                    messageBitIndex++;
                }
            }

            AudioInputStream waveWrite = new AudioInputStream(new ByteArrayInputStream(frames), format, frames.length / format.getFrameSize());

            AudioSystem.write(waveWrite, AudioFileFormat.Type.WAVE, new File(outputWavPath));
            System.out.println(45);
            waveWrite.close();


            System.out.println("Сообщение успешно спрятано в " + outputWavPath);

        } catch (FileNotFoundException e) {
            System.out.println("Ошибка: Один из файлов не найден.");
        } catch (IllegalArgumentException e) {
            System.out.println("Ошибка: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("Произошла непредвиденная ошибка: " + e.getMessage());
        }
    }

    public static String transpositionEncrypt(String message) {
        int lines = (int) Math.ceil((double) message.length() / 3);
        char[][] table = new char[lines][3];

        char[] textL = message.toCharArray();
        int index = 0;
        for (int i = 0; i < lines; i++) {
            for (int j = 0; j < 3; j++) {
                if (index < textL.length) {
                    table[i][j] = textL[index];
                    index++;
                } else {
                    table[i][j] = ' ';
                }
            }
        }

        StringBuilder ciphertext = new StringBuilder();
        for (int j = 0; j < 3; j++) {
            for (int i = 0; i < lines; i++) {
                ciphertext.append(table[i][j]);
            }
        }

        return ciphertext.toString();
    }

    public static String text(String messageFile) throws IOException {
        BufferedReader file = new BufferedReader(new InputStreamReader(new FileInputStream(messageFile), StandardCharsets.UTF_8));
        StringBuilder message = new StringBuilder();
        String line;
        while ((line = file.readLine()) != null) {
            message.append(line);
        }
        file.close();
        return message.toString();
    }

    public static void main(String[] args) {
        String inputWav = "C:\\Users\\nepip\\IdeaProjects\\encode\\src\\original.wav";//your file path
        String messageFile = "C:\\Users\\nepip\\IdeaProjects\\encode\\src\\message.txt";//your file path
        String outputWav = "C:\\Users\\nepip\\IdeaProjects\\encode\\src\\stego.wav";//your file path

        try {
            String textMessage = text(messageFile);
            String ciphertext = transpositionEncrypt(textMessage);
            hideMessageInWav(inputWav, ciphertext, outputWav);
        } catch (IOException e) {
            System.out.println("Ошибка: " + e.getMessage());
        }
    }
}

File: pythonCrypyionProject/encode.py
import wave
from math import ceil



def hide_message_in_wav(input_wav_path, message_file_path, output_wav_path):

    try:
        with wave.open(input_wav_path, 'rb') as wave_read:
            params = wave_read.getparams()
            frames = wave_read.readframes(params.nframes)
            frames = bytearray(frames)
            print(222)



        message_file_path += "====="
        message_bytes = message_file_path.encode('utf-8')

        if len(message_bytes) * 8 > len(frames):
            raise ValueError("Сообщение слишком длинное для этого аудиофайла.")

        message_bit_index = 0
        for i in range(len(message_bytes)):
            byte = message_bytes[i]
            for j in range(8):
                bit = (byte >> (7 - j)) & 1

                frames[message_bit_index] = (frames[message_bit_index] & 0xFE) | bit

                message_bit_index += 1

        with wave.open(output_wav_path, 'wb') as wave_write:
            wave_write.setparams(params)
            wave_write.writeframes(bytes(frames))

        print(f"Сообщение успешно спрятано в {output_wav_path}")

    except FileNotFoundError:
        print("Ошибка: Один из файлов не найден.")
    except ValueError as e:
        print(f"Ошибка: {e}")
    except Exception as e:
        print(f"Произошла непредвиденная ошибка: {e}")

def transposition_encrypt(message):

    lines=ceil(len(message)/3)

    table = [[' ' for _ in range(3)] for _ in range(lines)]

    text_l=list(message)

    index=0
    for i in range(lines):
        for j in range(3):
            table[i][j]=text_l[index]
            index+=1
            if index==len(text_l):
                break

    ciphertext = ''
    for j in range(3):
         for i in range(lines):
            ciphertext += table[i][j]

    return ciphertext

def text(message_file):
    with open(message_file, 'r', encoding='utf-8') as file:
        message = file.read()
    return message


if __name__ == "__main__":

    input_wav = "original.wav"
    message_file = "message.txt"
    output_wav = "stego.wav"

    text_message=text(message_file)
    ciphrotext=transposition_encrypt(text_message)
    hide_message_in_wav(input_wav, ciphrotext, output_wav)

